purchase_product_warn_message/models/purchase_order.py

```diff
@@ -0,0 +1,32 @@
+# Copyright 2024 Alberto Martínez <alberto.martinez@sygel.es>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class ModelName(models.Model):
+    _inherit = "purchase.order"
+
+    purchase_product_warn_msg = fields.Html(
+        compute="_compute_purchase_product_warn_msg"
+    )
+
+    @api.depends("state", "order_line.product_id")
+    def _compute_purchase_product_warn_msg(self):
+        for rec in self:
+            purchase_product_warn_msg = ""
+            separator = "<br/>"
+            if rec.order_line:
+                if rec.state in ["done", "cancel"]:
+                    rec.purchase_product_warn_msg = ""
+                    continue
+                warnable_products = rec.order_line.mapped("product_id").filtered(
+                    lambda p: p.purchase_line_warn == "warning"
+                )
+                for product_id in warnable_products:
+                    purchase_product_warn_msg += (
```

Comment:
You could create a list containing the warning messages of all the producst and them use separator.join() to join them, instead of checking wheter the current product is the last product.